File: cli/src/specsfy_cli/github.py
from __future__ import annotations

import os
import shutil
import subprocess


def api_headers(
    user_agent: str,
    *,
    accept: str = "application/vnd.github+json",
) -> dict[str, str]:
    headers = {
        "Accept": accept,
        "User-Agent": user_agent,
        "X-GitHub-Api-Version": "2022-11-28",
    }
    token = _token()
    if token:
        headers["Authorization"] = f"Bearer {token}"
    return headers


def _token() -> str | None:
    for variable in ("GH_TOKEN", "GITHUB_TOKEN"):
        token = os.environ.get(variable, "").strip()
        if token:
            return token

    executable = shutil.which("gh")
    if executable is None:
        return None
    result = subprocess.run(
        [executable, "auth", "token"],
        text=True,
        capture_output=True,
        check=False,
    )
    if result.returncode != 0:
        return None
    return result.stdout.strip() or None

File: cli/src/github.ts
/**
 * Autenticação compartilhada para chamadas à API do GitHub.
 */

import { execFile } from "node:child_process";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

/** Monta os headers oficiais e inclui token quando disponível. */
export async function apiHeaders(
  userAgent: string,
  accept = "application/vnd.github+json",
): Promise<Record<string, string>> {
  const headers: Record<string, string> = {
    Accept: accept,
    "User-Agent": userAgent,
    "X-GitHub-Api-Version": "2022-11-28",
  };
  const token = await githubToken();
  if (token) headers.Authorization = `Bearer ${token}`;
  return headers;
}

/** Resolve a credencial sem persistir nem imprimir seu valor. */
export async function githubToken(): Promise<string | undefined> {
  for (const variable of ["GH_TOKEN", "GITHUB_TOKEN"] as const) {
    const value = process.env[variable]?.trim();
    if (value) return value;
  }
  try {
    const result = await execFileAsync("gh", ["auth", "token"], {
      encoding: "utf8",
    });
    return result.stdout.trim() || undefined;
  } catch {
    return undefined;
  }
}
